File: appmonet-sdk/appmonet-sdk-mopub/src/main/java/com/monet/bidder/MopubNativeAdRequest.java
package com.monet.bidder;

import android.location.Location;
import android.os.Bundle;

import androidx.annotation.Nullable;

import com.monet.bidder.auction.AuctionRequest;
import com.monet.bidder.bid.BidResponse;
import com.mopub.nativeads.MoPubNative;
import com.mopub.nativeads.RequestParameters;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.Date;
import java.util.HashMap;
import java.util.Map;

import static com.monet.bidder.Constants.BIDS_KEY;
import static com.monet.bidder.Constants.Dfp.ADUNIT_KEYWORD_KEY;
import static com.monet.bidder.MoPubRequestUtil.getKeywords;
import static com.monet.bidder.MoPubRequestUtil.mergeKeywords;

class MopubNativeAdRequest extends AdServerAdRequest {
  private final Map<String, Object> mLocalExtras;
  private final String adUnitId;

  private MoPubNative moPubNative;
  private final RequestParameters requestParameters;
  RequestParameters modifiedRequestParameters;
  @Nullable
  private BidResponse mBid = null;

  MopubNativeAdRequest(MoPubNative moPubNative, String adUnitId,
      RequestParameters requestParameters) {
    this.requestParameters = requestParameters;
    mLocalExtras = new HashMap<>();
    this.moPubNative = moPubNative;
    this.adUnitId = adUnitId;
    if (mLocalExtras.containsKey(BIDS_KEY)) {
      try {
        mBid = BidResponse.Mapper.from(new JSONObject((String) mLocalExtras.get(BIDS_KEY)));
      } catch (JSONException e) {
        //do nothing
      }
    }
  }

  MopubNativeAdRequest(String adUnitId) {
    this.requestParameters = new RequestParameters.Builder().build();
    mLocalExtras = new HashMap<>();
    this.adUnitId = adUnitId;
  }

  static MopubNativeAdRequest fromAuctionRequest(AuctionRequest request) {
    MopubNativeAdRequest adRequest = new MopubNativeAdRequest(request.getAdUnitId());
    for (String key : request.getTargeting().keySet()) {
      Object targeting = request.getTargeting().get(key);
      if (targeting != null) {
        adRequest.mLocalExtras.put(key, targeting);
      }
    }

    if (request.getBid() != null) {
      adRequest.mBid = request.getBid();
    }

    return adRequest;
  }

  @Override
  public boolean hasBid() {
    return mBid != null;
  }

  @Override
  public BidResponse getBid() {
    return mBid;
  }

  @Override
  public Location getLocation() {
    if (requestParameters == null) {
      return null;
    }
    return requestParameters.getLocation();
  }

  @Override
  public Date getBirthday() {
    return null;
  }

  @Override
  public String getContentUrl() {
    return null;
  }

  @Override
  public String getGender() {
    return null;
  }

  @Override
  public AuctionRequest apply(AuctionRequest instance, AdServerAdView adView) {
    return instance;
  }

  @Override
  public Bundle getCustomTargeting() {
    return null;
  }

  void applyToView(MopubNativeAdView mopubNativeAdView) {
    // apply the targeting to the view, as keywords
    MoPubNative nativeAd = mopubNativeAdView.getMopubNative();
    mLocalExtras.put(BIDS_KEY, BidResponse.Mapper.toJson(mBid).toString());
    mLocalExtras.put(ADUNIT_KEYWORD_KEY, adUnitId);
    nativeAd.setLocalExtras(mLocalExtras);
    String keywords = getKeywords(mLocalExtras);
    if (requestParameters != null) {
      keywords = mergeKeywords(requestParameters.getKeywords(), keywords);
    }
    modifiedRequestParameters = setKeywords(keywords);
  }

  private RequestParameters setKeywords(String keywords) {
    if (requestParameters != null) {
      return new RequestParameters.Builder()
          .keywords(keywords)
          .location(requestParameters.getLocation())
          .userDataKeywords(requestParameters.getUserDataKeywords()).build();
    } else {
      return new RequestParameters.Builder()
          .keywords(keywords)
          .build();
    }
  }
}

File: appmonet-sdk/appmonet-sdk-mopub/src/main/java/com/monet/bidder/MopubNativeAdRequest.kt
package com.monet.bidder

import com.monet.AdServerAdRequest
import com.monet.AdServerAdView
import com.monet.BidResponse.Mapper.from
import com.monet.BidResponse.Mapper.toJsonString
import com.monet.LocationData
import com.monet.auction.AuctionRequest
import com.monet.bidder.Constants.BIDS_KEY
import com.monet.bidder.Constants.Dfp.ADUNIT_KEYWORD_KEY
import com.monet.bidder.MoPubRequestUtil.getKeywords
import com.monet.bidder.MoPubRequestUtil.mergeKeywords
import com.mopub.nativeads.MoPubNative
import com.mopub.nativeads.RequestParameters
import com.mopub.nativeads.RequestParameters.Builder
import org.json.JSONException
import java.util.HashMap

internal class MopubNativeAdRequest : AdServerAdRequest {
  private val mLocalExtras: MutableMap<String, Any>
  private val adUnitId: String
  private var moPubNative: MoPubNative? = null
  private val requestParameters: RequestParameters?
  var modifiedRequestParameters: RequestParameters? = null

  constructor(
    moPubNative: MoPubNative?,
    adUnitId: String,
    requestParameters: RequestParameters?
  ) {
    this.requestParameters = requestParameters
    mLocalExtras = mutableMapOf()
    this.moPubNative = moPubNative
    this.adUnitId = adUnitId
    mLocalExtras[BIDS_KEY]?.let {
      try {
        bid = from(it as String)
      } catch (e: JSONException) {
        //do nothing
      }
    }
  }

  constructor(adUnitId: String) {
    requestParameters = Builder().build()
    mLocalExtras = HashMap()
    this.adUnitId = adUnitId
  }

  override val location: LocationData?
    get() = requestParameters?.location?.let {
      LocationData(
          it.latitude, it.longitude, it.accuracy.toDouble(), it.provider
      )
    }

  override val birthday: Long?
    get() = null

  override val contentUrl: String?
    get() = null

  override val gender: String?
    get() = null

  override fun apply(
    request: AuctionRequest,
    adView: AdServerAdView
  ): AuctionRequest {
    return request
  }

  override val publisherProvidedId: String?
    get() = null

  override val customTargeting: Map<String, Any>
    get() = mapOf()

  fun applyToView(mopubNativeAdView: MopubNativeAdView) {
    // apply the targeting to the view, as keywords
    val nativeAd = mopubNativeAdView.mopubNative
    mLocalExtras[BIDS_KEY] = toJsonString(bid)
    mLocalExtras[ADUNIT_KEYWORD_KEY] = adUnitId
    nativeAd.setLocalExtras(mLocalExtras)
    var keywords = getKeywords(mLocalExtras)
    keywords = mergeKeywords(requestParameters?.keywords, keywords)
    modifiedRequestParameters = setKeywords(keywords)
  }

  private fun setKeywords(keywords: String): RequestParameters {
    return if (requestParameters != null) {
      Builder()
          .keywords(keywords)
          .location(requestParameters.location)
          .userDataKeywords(requestParameters.userDataKeywords).build()
    } else {
      Builder()
          .keywords(keywords)
          .build()
    }
  }

  companion object {
    fun fromAuctionRequest(request: AuctionRequest): MopubNativeAdRequest {
      val adRequest = MopubNativeAdRequest(request.adUnitId)
      for (key in request.targeting.keys) {
        val targeting = request.targeting[key]
        if (targeting != null) {
          adRequest.mLocalExtras[key] = targeting
        }
      }
      if (request.bid != null) {
        adRequest.bid = request.bid
      }
      return adRequest
    }
  }
}